extension/src-language-server/stpa/diagram/utils.ts

```diff
@@ -0,0 +1,224 @@
+ * @returns The number of the layer {@code node} should be in.
+ */
+function determineLayerForSTPANode(node: STPANode, hazardDepth: number, sysConsDepth: number, map: Map<string, number>, groupUCAs: groupValue): number {
+    switch (node.aspect) {
+        case STPAAspect.LOSS:
+            return 0;
+        case STPAAspect.HAZARD:
+            return 1 + node.hierarchyLvl;
+        case STPAAspect.SYSTEMCONSTRAINT:
+            return 2 + hazardDepth + node.hierarchyLvl;
+        case STPAAspect.RESPONSIBILITY:
+            return 3 + hazardDepth + sysConsDepth;
+        case STPAAspect.UCA:
+            // each UCA group gets its own layer
+            switch (groupUCAs) {
+                case groupValue.CONTROL_ACTION:
+                    if (node.controlAction && !map.has(node.controlAction)) {
+                        map.set(node.controlAction, map.size);
+                    }
+                    return 4 + hazardDepth + sysConsDepth + map.get(node.controlAction!)!;
+                case groupValue.SYSTEM_COMPONENT:
+                    if (node.controlAction && !map.has(node.controlAction.substring(0, node.controlAction.indexOf(".")))) {
+                        map.set(node.controlAction.substring(0, node.controlAction.indexOf(".")), map.size);
+                    }
+                    return 4 + hazardDepth + sysConsDepth + map.get(node.controlAction!.substring(0, node.controlAction!.indexOf(".")))!;
+                default:
+                    return 4 + hazardDepth + sysConsDepth;
+            }
+        case STPAAspect.CONTROLLERCONSTRAINT:
+            return 5 + hazardDepth + sysConsDepth + map.size;
+        case STPAAspect.SCENARIO:
+            return 6 + hazardDepth + sysConsDepth + map.size;
+        case STPAAspect.SAFETYREQUIREMENT:
+            return 7 + hazardDepth + sysConsDepth + map.size;
+        default:
+            return -1;
+    }
+}
+
+/**
+ * Sets the level property for {@code nodes} depending on the layer they should be in.
+ * @param nodes The nodes representing the stpa components.
+ * @param groupUCAs Determines whether and how UCAs are grouped.
+ */
+export function setLevelsForSTPANodes(nodes: STPANode[], groupUCAs: groupValue): void {
+    // determines the maximal hierarchy depth of hazards and system constraints
+    let maxHazardDepth = -1;
+    let maxSysConsDepth = -1;
+    for (const node of nodes) {
+        if (node.aspect === STPAAspect.HAZARD) {
+            maxHazardDepth = maxHazardDepth > node.hierarchyLvl ? maxHazardDepth : node.hierarchyLvl;
+        }
+        if (node.aspect === STPAAspect.SYSTEMCONSTRAINT) {
+            maxSysConsDepth = maxSysConsDepth > node.hierarchyLvl ? maxSysConsDepth : node.hierarchyLvl;
+        }
+    }
+
+    // used to determine which control action or system component belongs to which group number
+    const map = new Map<string, number>();
+    // sets level property to the layer of the nodes.
```

Comment:
this comment doesn't explain anything, what does the level property represent in contrast to the layer? since it's being set to -layer